src/hotspot/share/gc/g1/g1RemSet.cpp

```diff
@@ -488,17 +490,122 @@ G1RemSet::G1RemSet(G1CollectedHeap* g1h,
+    G1CollectedHeap::heap()->collection_set()->update_young_region_prediction(r, rs_length);
+
+    _regions_visited++;
+
+    if (_regions_visited == 10) {
+      if (_sts->should_yield()) {
+        _sts->yield();
+        // A gc may have occurred and our sampling data is stale and further
+        // traversal of the collection set is unsafe
+        return true;
+      }
+      _regions_visited = 0;
+    }
+    return false;
+  }
+
+  size_t sampled_rs_length() const { return _sampled_rs_length; }
+};
+
+// Task handling young gen remembered set sampling.
+class G1RemSetSamplingTask : public G1ServiceTask {
+  // Sample the current length of remembered sets for young.
+  //
+  // At the end of the GC G1 determines the length of the young gen based on
+  // how much time the next GC can take, and when the next GC may occur
+  // according to the MMU.
+  //
+  // The assumption is that a significant part of the GC is spent on scanning
+  // the remembered sets (and many other components), so this thread constantly
+  // reevaluates the prediction for the remembered set scanning costs, and potentially
+  // G1Policy resizes the young gen. This may do a premature GC or even
+  // increase the young gen size to keep pause time length goal.
+  void sample_young_list_rs_length(SuspendibleThreadSetJoiner* sts){
+    G1CollectedHeap* g1h = G1CollectedHeap::heap();
+    G1Policy* policy = g1h->policy();
+
+    if (policy->use_adaptive_young_list_length()) {
+      G1YoungRemSetSamplingClosure cl(sts);
+
+      G1CollectionSet* g1cs = g1h->collection_set();
+      g1cs->iterate(&cl);
+
+      if (cl.is_complete()) {
+        policy->revise_young_list_target_length_if_necessary(cl.sampled_rs_length());
+      }
+    }
+  }
+
+  // To avoid extensive rescheduling if the task is executed a bit early. The task is
+  // only rescheduled if the expected time is more than 1ms away.
+  bool should_reschedule() {
+    return reschedule_delay_ms() > 1;
+  }
+
+  // There is no reason to do the sampling if a GC occurred recently. We use the
+  // G1ConcRefinementServiceIntervalMillis as the metric for recently and calculate
+  // the diff to the last GC. If the last GC occurred longer ago than the interval
+  // 0 is returned.
+  jlong reschedule_delay_ms() {
+    Tickspan since_last_gc = G1CollectedHeap::heap()->time_since_last_collection();
+    jlong delay = (jlong) (G1ConcRefinementServiceIntervalMillis - since_last_gc.milliseconds());
+    return MAX2<jlong>(0L, delay);
+  }
+
+public:
+  G1RemSetSamplingTask(const char* name) : G1ServiceTask(name) { }
+  virtual void execute() {
+    SuspendibleThreadSetJoiner sts;
+
+    // Reschedule if a GC happened too recently.
+    if (should_reschedule()) {
+      // Calculate the delay given the last GC and the interval.
+      schedule(reschedule_delay_ms());
+      return;
+    }
+
+    // Do the actual sampling.
+    sample_young_list_rs_length(&sts);
+    schedule(G1ConcRefinementServiceIntervalMillis);
```

Comment:
I am good with the new location of the code but I would prefer if this change would be limited to the move and not change the rescheduling logic.
If it's not too much of a bother please split this out into a second CR.

Reply:
Thanks for looking at this Thomas. I removed the scheduling logic and filed: [JDK-8252645](https://bugs.openjdk.java.net/browse/JDK-8252645)